[pulsar-client-cpp] New port

BewareMyPower opened the pull request:
- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [x] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [x] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html)
- [x] The versioning scheme in `vcpkg.json` matches what upstream says.
- [x] The license declaration in `vcpkg.json` matches what upstream says.
- [x] The installed as the "copyright" file matches what upstream says.
- [x] The source code of the component installed comes from an authoritative source.
- [x] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is in the new port's versions file.
- [x] Only one version is added to each modified port's versions file.

Regarding the patch
1. The upstream uses lots of raw `find_path`/`find_libraries` calls to add a path of a library to the `COMMON_LIBS` list, this patch replaces all of them with the `find_package` calls and adds the target to the `COMMON_LIBS` list.
2. The upstream allows user passing a `PROTOC_PATH`  or find it from the system path, then runs a shell command to generate the protobuf source files. This patch uses the `protobuf_generate` function instead.

BewareMyPower committed:
[pulsar-client-cpp] New port

BewareMyPower commented:
@microsoft-github-policy-service agree

BewareMyPower commented:
Is there anyway I can check the error logs in CI? e.g. the error logs on x64_linux is `/mnt/vcpkg-ci/buildtrees/pulsar-client-cpp/install-x64-linux-dbg-out.log` but I cannot find a way to download it.

dg0yt commented:
> Is there anyway I can check the error logs in CI? e.g. the error logs on x64_linux is `/mnt/vcpkg-ci/buildtrees/pulsar-client-cpp/install-x64-linux-dbg-out.log` but I cannot find a way to download it.

https://github.com/microsoft/vcpkg/discussions/31357#discussioncomment-5855044

dg0yt commented:
If the cmake config is not provided/merged by upstream, it has to be prefixed with `unofficial`.
https://learn.microsoft.com/en-us/vcpkg/contributing/maintainer-guide#add-cmake-exports-in-an-unofficial--namespace

Learn to use `find_package()` /  `find_dependency()`. For static linkage, you need all libs transitively, on non-MSVC plaforms also in the right order. It is not maintainable with this `find_library()` approach.

BewareMyPower commented:
I see. I will fix it soon.

BewareMyPower commented:
I renamed the `.cmake` file and updated the target name to mark them as unofficial. For 3rd party dependencies, I used `find_package` now.

Regarding the Windows and Linux build failures, they seem related to the official way to find `protoc`. I'm trying to add a patch for it.

BewareMyPower committed:
Remove unnecessary link options

BewareMyPower commented:
@jimwang118 Now all tests passed, could you take a look?

jimwang118 commented:
Usage test pass with following triplets:

```
x64-windows
x64-windows-static
```

vicroms commented on ports/pulsar-client-cpp/unofficial-pulsar-config.cmake:
```diff
@@ -0,0 +1,37 @@
+        message(FATAL_ERROR "Broken installation of vcpkg port pulsar-client-cpp")
+    endif ()
+
+    find_package(protobuf CONFIG REQUIRED)
+    find_package(CURL CONFIG REQUIRED)
+    find_package(zstd CONFIG REQUIRED)
+    find_package(snappy CONFIG REQUIRED)
```
```suggestion
    include(CMakeFindDependencyMacro)
    find_dependency(protobuf)
    find_dependency(CURL)
    find_dependency(zstd)
    find_dependency(snappy)
```

This should use `find_dependency` instead

vicroms commented on ports/pulsar-client-cpp/portfile.cmake:
```diff
@@ -0,0 +1,35 @@
+      0001-use-find-package.patch
+)
+
+string(COMPARE EQUAL "${VCPKG_LIBRARY_LINKAGE}" "static" PULSAR_BUILD_STATIC)
```
Is this really `ONLY_STATIC_LIBRARY`? Having to set `PULSAR_BUILD_STATIC` seems to suggest that this may be able to build dynamically.

BewareMyPower replied:
Good suggestion. I will use `PULSAR_ONLY_STATIC_LIBRARY` as the name.

BewareMyPower replied:
Sorry I misunderstood it. I changed the name to `BUILD_STATIC_LIB` now. Since the original repo defines two CMake variables `BUILD_DYNAMIC_LIB` and `BUILD_STATIC_LIB` that indicate whether to build the dynamic library and the static library. See https://github.com/apache/pulsar-client-cpp/blob/7087a8ef730a1422980f5ee01399dc3528eebe10/CMakeLists.txt#L68-L72

So I'm going to add these options explicitly:

```cmake
string(COMPARE EQUAL "${VCPKG_LIBRARY_LINKAGE}" "static" BUILD_STATIC_LIB)

vcpkg_cmake_configure(
# ...
        -DBUILD_DYNAMIC_LIB=ON
        -DBUILD_STATIC_LIB=${BUILD_STATIC_LIB}
)
```

so that it would be clear that:
- When the linkage is static, build both dynamic and static libraries
- Otherwise, build only the dynamic library

jimwang118 replied:
When I tested x64-windows locally, the VCPKG_LIBRARY_LINKAGE link status was dynamic, but only static libraries were generated, not dynamic libraries.

BewareMyPower replied:
Got it, I will take a look soon.

BewareMyPower committed:
Address review comments

- Use find_dependency instead of find_package
- Specify BUILD_DYNAMIC_LIB explicitly

jimwang118 commented:
@vicroms Is there anything else that needs to be modified in this PR?

BewareMyPower committed:
Fix the Windows build and linkage

BewareMyPower committed:
Merge branch 'master' into add-pulsar-port-complete

BewareMyPower committed:
Fix dynamic library not built when VCPKG_LIBRARY_LINKAGE is dynamic on Windows

BewareMyPower committed:
Fix Linux and OSX failures

BewareMyPower commented:
Now I changed the build strategy when `VCPKG_LIBRARY_LINKAGE` is static:
- On Linux or Darwin, build both the static library and dynamic library. By default, the static library is linked. Users can enable `PULSAR_FORCE_DYNAMIC_LIBRARY` to link the dynamic library.
- On Windows, build the static library only.

BTW, I just tested it on Windows for both `x64-windows` and `x64-windows-static` build. It seems that the CI workflow does not verify if the imported CMake target is correct.

Here are the steps for reference: https://gist.github.com/BewareMyPower/fa81e6f8badb494ba195ef5273a4e4db

dg0yt commented:
> * On Linux or Darwin, build both the static library and dynamic library. By default, the static library is linked. Users can enable `PULSAR_FORCE_DYNAMIC_LIBRARY` to link the dynamic library.

Why doesn't it simply follow the triplet?

BewareMyPower commented:
> Why doesn't it simply follow the triplet (or actually VCPKG_LIBRARY_LINKAGE)?

See https://github.com/microsoft/vcpkg/discussions/35733

And it's yes that we can follow the triplet because we can customize the triplet to achieve my goal. Therefore, I will remove this complicated CMake option.

BewareMyPower committed:
Remove PULSAR_FORCE_DYNAMIC_LIBRARY and upgrade version to 3.4.2

dg0yt reviewed (commented):
IMO the patch should be less invasive.
(Note that you can't apply the PR suggestions directly. I didn't adjust the line counts.)

You are already patching the CMake build system. Then another patch should take care of  really exporting the desired targets from the build system. (The main config file is still needed to do the `find_dependency` stuff for resolving targets in interface properties.)

BewareMyPower commented on ports/pulsar-client-cpp/0001-use-find-package.patch:
```diff
@@ -0,0 +1,154 @@
+
+ add_definitions(-DHAS_SNAPPY=${HAS_SNAPPY})
+
+-if(NOT APPLE AND NOT MSVC)
+-    set(COMMON_LIBS ${COMMON_LIBS} rt)
+-endif ()
+-
```
Oh, this change was made because `rt` cannot be found on Android platforms. But later I found it's hard to adapt the library to Android so I removed the support for Android. We can revert this change.

BewareMyPower commented on ports/pulsar-client-cpp/0001-use-find-package.patch:
```diff
@@ -0,0 +1,154 @@
+
+-# Protobuf generation is only supported natively starting from CMake 3.8
+-# Using custom command for now
++add_library(proto-objects OBJECT "${CMAKE_SOURCE_DIR}/proto/PulsarApi.proto")
++target_link_libraries(proto-objects PUBLIC protobuf::libprotobuf)
++target_include_directories(proto-objects PUBLIC "${LIB_AUTOGEN_DIR}")
++protobuf_generate(
++    TARGET proto-objects
++    IMPORT_DIRS "${CMAKE_SOURCE_DIR}/proto"
++    PROTOC_OUT_DIR "${LIB_AUTOGEN_DIR}")
++
+ set(PROTO_SOURCES ${LIB_AUTOGEN_DIR}/PulsarApi.pb.cc ${LIB_AUTOGEN_DIR}/PulsarApi.pb.h)
+ set(PULSAR_SOURCES ${PULSAR_SOURCES} ${PROTO_SOURCES})
+-ADD_CUSTOM_COMMAND(
+-         OUTPUT ${PROTO_SOURCES}
+-         COMMAND ${PROTOC_PATH} -I ../proto ../proto/PulsarApi.proto --cpp_out=${LIB_AUTOGEN_DIR}
+-         DEPENDS
+-         ../proto/PulsarApi.proto
+-         WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR})
```
Yes.

BewareMyPower committed:
Reduce the changes to the upstream CMakeLists.txt

BewareMyPower commented on ports/pulsar-client-cpp/0001-use-find-package.patch:
```diff
@@ -0,0 +1,154 @@
+-    find_package(Protobuf REQUIRED)
+-endif ()
++find_package(protobuf REQUIRED CONFIG)
++set(COMMON_LIBS ${COMMON_LIBS} protobuf::libprotobuf)
```
Why not just use `find_package(protobuf REQUIRED)` here?

BillyONeal commented:
Sorry I broke the build lab Sunday night

jimwang118 commented on versions/p-/pulsar-client-cpp.json:
```diff
@@ -0,0 +1,14 @@
+      "version": "3.4.2",
+      "port-version": 0
+    },
+    {
+      "git-tree": "f0e8879f5638f152daf0a97e9a16707f89f93f9e",
+      "version": "3.4.1",
+      "port-version": 0
+    }
+  ]
+}
```
```suggestion
```

jimwang118 replied:
If you are starting from version 3.4.2, it is recommended to delete the version record of 3.4.1.

jimwang118 commented:
Note: I will be converting your PR to draft status. The above suggested changes are only recommendations. If you are willing to adopt them, please click "ready for review" after making the modifications. If you do not wish to make any changes, please click "ready for review" directly. That way, I can be aware that you've responded since you can't modify the tags.

BewareMyPower committed:
Remove unused version

dg0yt commented on ports/pulsar-client-cpp/unofficial-pulsar-config.cmake:
```diff
@@ -0,0 +1,58 @@
+    get_filename_component(VCPKG_IMPORT_PREFIX "${VCPKG_IMPORT_PREFIX}" PATH)
+    get_filename_component(VCPKG_IMPORT_PREFIX "${VCPKG_IMPORT_PREFIX}" PATH)
+
+    find_path(_pulsar_include_dir NAMES "pulsar/Client.h" PATH ${VCPKG_IMPORT_PREFIX})
```
```suggestion
    find_path(_pulsar_include_dir NAMES "pulsar/Client.h" PATH "${VCPKG_IMPORT_PREFIX}/include" REQUIRED)
```
BTW which style is officially to be used used:
~~~
include <pulsar/Client.h>
vs.
include <Client.h>
~~~

BewareMyPower replied:
`#include <pulsar/Client.h>` is recommended.

dg0yt commented on ports/pulsar-client-cpp/unofficial-pulsar-config.cmake:
```diff
@@ -0,0 +1,58 @@
+    get_filename_component(VCPKG_IMPORT_PREFIX "${VCPKG_IMPORT_PREFIX}" PATH)
+
+    find_path(_pulsar_include_dir NAMES "pulsar/Client.h" PATH ${VCPKG_IMPORT_PREFIX})
+    if (CMAKE_BUILD_TYPE STREQUAL "Debug")
+        set(VCPKG_IMPORT_PREFIX "${VCPKG_IMPORT_PREFIX}/debug")
+    endif ()
```
This won't work for multi-config generators like Visual Studio.

And it also won't work for usage of single-triplet (release) installations in a debug build.

BewareMyPower replied:
I will handle this issue later on my Windows laptop. For now, I have a question for:

>  single-triplet (release) installations in a debug build.

Did you mean in this case, a release library (e.g. `bin/*.dll` rather than `debug/bin/*.dll`) is linked for a debug build?

dg0yt commented on ports/pulsar-client-cpp/unofficial-pulsar-config.cmake:
```diff
@@ -0,0 +1,58 @@
+    if (CMAKE_BUILD_TYPE STREQUAL "Debug")
+        set(VCPKG_IMPORT_PREFIX "${VCPKG_IMPORT_PREFIX}/debug")
+    endif ()
+    if ("@VCPKG_LIBRARY_LINKAGE@" STREQUAL "dynamic")
+        find_library(_pulsar_library NAMES libpulsar.so libpulsar.dylib pulsar.lib PATH ${VCPKG_IMPORT_PREFIX})
+        set(_pulsar_link_static_library OFF)
+        message(STATUS "Found dynamic _pulsar_library: ${_pulsar_library}")
+    else ()
+        find_library(_pulsar_library NAMES libpulsar.a pulsar-static.lib PATH ${VCPKG_IMPORT_PREFIX})
+        set(_pulsar_link_static_library ON)
+        message(STATUS "Found static _pulsar_library: ${_pulsar_library}")
+    endif ()
```
```suggestion
    find_library(_pulsar_library NAMES pulsar PATH "${VCPKG_IMPORT_PREFIX}" PATH_SUFFIXES "lib")
```

dg0yt commented on ports/pulsar-client-cpp/unofficial-pulsar-config.cmake:
```diff
@@ -0,0 +1,58 @@
+        find_dependency(dlfcn-win32 CONFIG)
+    endif ()
+
+    add_library(unofficial::pulsar::pulsar INTERFACE IMPORTED)
```
```suggestion
    add_library(unofficial::pulsar::pulsar UNKNOWN IMPORTED)
```
Interface libs don't have a location.

dg0yt commented on ports/pulsar-client-cpp/unofficial-pulsar-config.cmake:
```diff
@@ -0,0 +1,58 @@
+        INTERFACE_INCLUDE_DIRECTORIES "${_pulsar_include_dir}"
+        IMPORTED_LOCATION "${_pulsar_library}")
+    set(DEPENDENCIES
+        ${_pulsar_library}
```
```suggestion
```
after properly declaring it as unknown type.

dg0yt commented on ports/pulsar-client-cpp/unofficial-pulsar-config.cmake:
```diff
@@ -0,0 +1,58 @@
+    if (_pulsar_link_static_library)
+        target_compile_definitions(unofficial::pulsar::pulsar INTERFACE PULSAR_STATIC)
+    endif ()
+    target_link_libraries(unofficial::pulsar::pulsar INTERFACE ${DEPENDENCIES})
```
Just add them directly, without storing them in `DEPENDENCIES` first?

BewareMyPower replied:
I added `DEPENDENCIES` because the `dlfcn-win32::dl` dependency is only for Windows. I have thought the 2nd `target_link_libraries` call will overwrite the 1st call.

dg0yt replied:
`target_link_libraries` doesn't overwrite. It is the normal command to add link libraries.
(In older versions of CMake it couldn't operate on imported targets, and then scripts would need to initialize the target properties directly. But even in that case, there is one command to append to properties instead of overwriting.)

dg0yt commented on ports/pulsar-client-cpp/unofficial-pulsar-config.cmake:
```diff
@@ -0,0 +1,58 @@
+    if (MSVC)
+        set(DEPENDENCIES ${DEPENDENCIES} dlfcn-win32::dl)
+    endif ()
+    if (_pulsar_link_static_library)
```
```suggestion
    if (if ("@VCPKG_LIBRARY_LINKAGE@" STREQUAL "static"))
```

dg0yt commented on ports/pulsar-client-cpp/unofficial-pulsar-config.cmake:
```diff
@@ -0,0 +1,58 @@
+        target_compile_definitions(unofficial::pulsar::pulsar INTERFACE PULSAR_STATIC)
+    endif ()
+    target_link_libraries(unofficial::pulsar::pulsar INTERFACE ${DEPENDENCIES})
+    unset(_pulsar_link_static_library)
+    unset(_pulsar_library)
+    unset(_pulsar_include_dir)
```
The first would become obsolete with my proposal.
The second and third as cache variables and cannot be unset this way.

dg0yt commented on ports/pulsar-client-cpp/unofficial-pulsar-config.cmake:
```diff
@@ -0,0 +1,58 @@
+        set(DEPENDENCIES ${DEPENDENCIES} dlfcn-win32::dl)
+    endif ()
+    if (_pulsar_link_static_library)
+        target_compile_definitions(unofficial::pulsar::pulsar INTERFACE PULSAR_STATIC)
```
For the poor msbuild project without cmake, it would be better to fixup the installed headers so that this definition is no longer needed.

dg0yt reviewed (commented):
I still think it would be better to export the cmake targets. This immediately solves the problems for multi-config generators.

BewareMyPower committed:
Optimize finding and linking dependency and patch the header for static library

BewareMyPower committed:
Support multi-config generators

BewareMyPower committed:
Fix path for release libraries

BewareMyPower commented:
Hi @dg0yt I have addressed your comments and tested with `x64-windows` and `x64-windows-static` triplets for both debug and release builds, PTAL again. 

BTW, I found the result of `find_library` varies for the `CMAKE_BUILD_TYPE` variable. For example, when it's `Debug`, the result will be `debug/lib/pulsar.lib` or `debug/lib/pulsar-static.lib`. When it's `Release`, the result becomes `lib/pulsar.lib` or `lib/pulsar-static.lib`. So I used `CMAKE_IGNORE_PATH` to ignore the `debug/` path. Not sure if there is a better solution.

BewareMyPower commented:
ping @dg0yt @jimwang118 @vicroms

BewareMyPower commented:
It has been pending for over a week. Could anyone take a look?
